test/extensions/common/aws/utility_test.cc

```diff
@@ -141,6 +144,144 @@ content-hash)",
+  std::string path = "/test//test///";
+  const auto canonical_path = Utility::canonicalizePathString(path, "s3");
+  EXPECT_EQ("/test//test///", canonical_path);
+}
+
+TEST(UtilityTest, CanonicalizeS3PathStringSlashes) {
+  std::string path = "///";
+  const auto canonical_path = Utility::canonicalizePathString(path, "s3");
+  EXPECT_EQ("///", canonical_path);
+}
+
+TEST(UtilityTest, CanonicalizeS3PathStringWithEncoding) {
+  std::string path = "/test$file.txt";
+  const auto canonical_path = Utility::canonicalizePathString(path, "s3");
+  EXPECT_EQ("/test%24file.txt", canonical_path);
+}
+
+TEST(UtilityTest, CanonicalizeS3PathStringWithEncodingSpaces) {
+  std::string path = "/test and test/";
+  const auto canonical_path = Utility::canonicalizePathString(path, "s3");
+  EXPECT_EQ("/test%20and%20test/", canonical_path);
+}
+
+TEST(UtilityTest, EncodePathSegment) {
+  std::string path = "test^!@=-_~.";
+  const auto encoded_path = Utility::encodePathSegment(path, "appmesh");
+  EXPECT_EQ("test%5E%21%40%3D-_~.", encoded_path);
+}
+
+TEST(UtilityTest, EncodeS3PathSegment) {
+  std::string path = "/test/^!@=/-_~.";
+  const auto encoded_path = Utility::encodePathSegment(path, "s3");
+  EXPECT_EQ("/test/%5E%21%40%3D/-_~.", encoded_path);
+}
+
+TEST(UtilityTest, CanonicalizeQueryString) {
+  std::string query = "a=1&b=2";
+  const auto canonical_query = Utility::canonicalizeQueryString(query);
+  EXPECT_EQ("a=1&b=2", canonical_query);
+}
+
+TEST(UtilityTest, CanonicalizeQueryStringTrailingEquals) {
+  std::string query = "a&b";
+  const auto canonical_query = Utility::canonicalizeQueryString(query);
+  EXPECT_EQ("a=&b=", canonical_query);
+}
+
+TEST(UtilityTest, CanonicalizeQueryStringSorted) {
+  std::string query = "a=3&b=1&a=2&a=1";
+  const auto canonical_query = Utility::canonicalizeQueryString(query);
+  EXPECT_EQ("a=1&a=2&a=3&b=1", canonical_query);
+}
+
+TEST(UtilityTest, CanonicalizeQueryStringEncoded) {
+  std::string query = "a=^!@&b=/-_~.";
+  const auto canonical_query = Utility::canonicalizeQueryString(query);
+  EXPECT_EQ("a=%5E%21%40&b=%2F-_~.", canonical_query);
+}
+
+TEST(UtilityTest, CanonicalizeQueryStringWithPlus) {
+  std::string query = "a=1+2";
+  const auto canonical_query = Utility::canonicalizeQueryString(query);
+  EXPECT_EQ("a=1%202", canonical_query);
+}
+
+TEST(UtilityTest, CanonicalizeQueryStringDoubleEncodeEquals) {
+  std::string query = "a=!.!=!";
+  const auto canonical_query = Utility::canonicalizeQueryString(query);
+  EXPECT_EQ("a=%21.%21%253D%21", canonical_query);
+}
+
+TEST(UtilityTest, EncodeQuerySegment) {
+  std::string query = "^!@/-_~.";
+  const auto encoded_query = Utility::encodeQueryParam(query);
+  EXPECT_EQ("%5E%21%40%2F-_~.", encoded_query);
+}
+
+TEST(UtilityTest, EncodeQuerySegmentReserved) {
+  std::string query = "?=&";
+  const auto encoded_query = Utility::encodeQueryParam(query);
+  EXPECT_EQ("%3F%253D%26", encoded_query);
+}
+
```

Comment:
Can you add a test that loops over every possible 3-character query and simply just calls `encodeQueryParam`, `encodePathSegment` and `canonicalizeQueryString`? It should only take a couple seconds to run and it might catch any weird cases that would cause a segfault.

Reply:
Add this fuzz test, but it takes about 40 seconds to run though.

Reply:
Update: It takes about 9 secs to run on my box. And tsan test takes about 78 secs.